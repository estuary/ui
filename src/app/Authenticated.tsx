import { Skeleton } from '@mui/material';
import AppLayout from 'AppLayout';
import PageNotFound from 'pages/error/PageNotFound';
import Home from 'pages/Home';
import { lazy, Suspense } from 'react';
import { Route, Routes } from 'react-router';

const Admin = lazy(() => import('../pages/Admin'));

const Builds = lazy(() => import('../pages/Builds'));

const Captures = lazy(() => import('../pages/Captures'));
const CaptureCreate = lazy(() => import('components/capture/create/index'));
const CaptureEdit = lazy(() => import('components/capture/edit/index'));

const Materializations = lazy(() => import('../pages/Materializations'));
const MaterializationCreate = lazy(
    () => import('../components/materialization/create')
);

const Collections = lazy(() => import('../pages/Collections'));

const Authenticated = () => {
    return (
        <Suspense fallback={<Skeleton animation="wave" />}>
            <Routes>
                <Route element={<AppLayout />}>
                    <Route path="/" element={<Home />} />
                    <Route path="/test/builds" element={<Builds />} />
                    <Route path="/dashboard" element={<Home />} />
                    <Route path="collections" element={<Collections />} />
                    <Route path="captures" element={<Captures />} />
                    <Route path="capture">
                        <Route path="create" element={<CaptureCreate />} />
                        <Route path="edit" element={<CaptureEdit />} />
                    </Route>

                    <Route
                        path="materializations"
                        element={<Materializations />}
                    />
                    <Route path="materialization">
                        <Route
                            path="create"
                            element={<MaterializationCreate />}
                        />
                    </Route>
                    <Route path="admin/*" element={<Admin />} />
                    <Route path="builds" element={<Builds />} />
                </Route>
                <Route path="*" element={<PageNotFound />} />
            </Routes>
        </Suspense>
    );
};

export default Authenticated;
