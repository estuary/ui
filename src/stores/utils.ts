import { isEmpty, map } from 'lodash';

export const checkForErrors = (obj: any) =>
    obj?.errors && obj.errors.length > 0;

export const populateErrors = (configDictionary: any) => {
    let configErrors: any[] = [];

    // TODO (errors) When there are not configs do we need to populate this object with something?
    if (Object.keys(configDictionary).length > 0) {
        map(configDictionary, (config) => {
            if (checkForErrors(config)) {
                configErrors = configErrors.concat(config.errors);
            }
        });
    }

    return {
        configErrors,
        hasErrors: !isEmpty(configErrors),
    };
};
