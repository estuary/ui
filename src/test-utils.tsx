// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react';
import { ReactElement, type FC } from 'react';
import { act } from 'react-dom/test-utils';
import { localStorageKey } from './auth';
import AppProviders from './context';

const AllTheProviders: FC = ({ children }) => {
    return <AppProviders>{children}</AppProviders>;
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

const loginAsUser = (userName: string = 'fakeUserName') => {
    act(() => {
        window.localStorage.setItem(localStorageKey, userName);
    });
};

export * from '@testing-library/react';
export { customRender as render, loginAsUser };
