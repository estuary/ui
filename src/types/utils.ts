export type WithRequiredProperty<T, K extends keyof T> = T & {
    [P in K]-?: T[P];
};

export type WithRequiredNonNullProperty<T, K extends keyof T> = T & {
    [P in K]-?: Exclude<T[P], null>;
};
