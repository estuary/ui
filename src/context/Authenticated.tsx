import { Auth } from '@supabase/ui';
import AppGuards from 'app/guards';
import { unauthenticatedRoutes } from 'app/routes';
import ConfirmationModalContextProvider from 'context/Confirmation';
import { ZustandProvider } from 'context/Zustand/provider';
import useLoginRedirectPath from 'hooks/useLoginRedirectPath';
import { Navigate, useLocation } from 'react-router-dom';
import { BaseComponentProps } from 'types';
import AuthEvents from './AuthEvents';
import PreFetchDataProvider from './fetcher';
import Osano from './Osano';
import QueryParamProvider from './QueryParam';

interface Props extends BaseComponentProps {
    firstLoad?: boolean;
}

// TODO (routes) not totaly sure where we want this kinda thing to live
//  in the long term so leaving it here for now.
export function RequireAuth({ children, firstLoad }: Props) {
    const { user } = Auth.useUser();
    const location = useLocation();
    const redirectTo = useLoginRedirectPath();

    if (user && firstLoad) {
        // Redirect to the welcome page if landed on the login page while logged in
        return <Navigate to={redirectTo} replace />;
    }

    if (!user && !firstLoad) {
        // Redirect to login with the current URL requested if no user
        return (
            <Navigate
                to={unauthenticatedRoutes.login.path}
                state={{ from: location }}
                replace
            />
        );
    }

    // eslint-disable-next-line react/jsx-no-useless-fragment
    return <>{children}</>;
}

// This is for contexts that should only be added to the app after the user has authenticated
export function AuthenticatedOnlyContext({ children }: BaseComponentProps) {
    return (
        <RequireAuth>
            <Osano>
                <QueryParamProvider>
                    <AppGuards>
                        <AuthEvents>
                            <PreFetchDataProvider>
                                <ZustandProvider>
                                    <ConfirmationModalContextProvider>
                                        {children}
                                    </ConfirmationModalContextProvider>
                                </ZustandProvider>
                            </PreFetchDataProvider>
                        </AuthEvents>
                    </AppGuards>
                </QueryParamProvider>
            </Osano>
        </RequireAuth>
    );
}
