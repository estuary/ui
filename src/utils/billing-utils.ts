import { isEqual, parseISO } from 'date-fns';
import { isEmpty, sum } from 'lodash';
import prettyBytes from 'pretty-bytes';
import { BillingRecord } from 'stores/Billing/types';
import { CatalogStats_Billing, Entity } from 'types';

export const TOTAL_CARD_HEIGHT = 300;

// Grid item height - 72 = graph canvas height
export const CARD_AREA_HEIGHT = TOTAL_CARD_HEIGHT - 72;

export const BYTES_PER_GB = 1073741824;

const FREE_TASK_COUNT = 2;

export enum FREE_GB_BY_TIER {
    FREE = 20,
    PERSONAL = 10,
    ENTERPRISE = 10,
}

export const evaluateSpecType = (query: CatalogStats_Billing): Entity => {
    if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
        return 'capture';
    } else if (Object.hasOwn(query.flow_document.taskStats, 'materialize')) {
        return 'materialization';
    } else {
        return 'collection';
    }
};

const evaluateTotalCost = (dataVolume: number, taskCount: number) => {
    const freeBytes = 10 * BYTES_PER_GB;

    const dataVolumeOverLimit =
        dataVolume > freeBytes ? dataVolume - freeBytes : 0;

    const taskCountOverLimit =
        taskCount > FREE_TASK_COUNT ? taskCount - FREE_TASK_COUNT : 0;

    return (
        (dataVolumeOverLimit / BYTES_PER_GB) * 0.75 + taskCountOverLimit * 20
    );
};

const evaluateDataVolume = (stats: CatalogStats_Billing[]): number => {
    const taskBytes: number[] = stats.map((query) => {
        if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
            return query.bytes_written_by_me;
        } else if (
            Object.hasOwn(query.flow_document.taskStats, 'materialize')
        ) {
            return query.bytes_read_by_me;
        } else {
            return query.bytes_written_by_me + query.bytes_read_by_me;
        }
    });

    return sum(taskBytes);
};

export const stripTimeFromDate = (date: string) => {
    const [truncatedDateStr] = date.split('T', 1);

    return parseISO(truncatedDateStr);
};

const getInitialBillingRecord = (date: string): BillingRecord => {
    const truncatedDate = stripTimeFromDate(date);

    return {
        date: truncatedDate,
        dataVolume: 0,
        taskCount: 0,
        totalCost: 0,
        pricingTier: null,
        taskRate: null,
        gbFree: null,
        includedTasks: null,
    };
};

// TODO (billing): Remove this helper function to translate data returned from
//   the new RPC when available.
export const formatBillingCatalogStats = (
    value: CatalogStats_Billing[]
): BillingRecord[] => {
    const taskStatData = value.filter((query) =>
        Object.hasOwn(query.flow_document, 'taskStats')
    );

    let sortedStats: {
        [date: string]: CatalogStats_Billing[];
    } = {};

    taskStatData.forEach((query) => {
        if (Object.hasOwn(sortedStats, query.ts)) {
            sortedStats[query.ts].push(query);
        } else {
            sortedStats = {
                ...sortedStats,
                [query.ts]: [query],
            };
        }
    });

    const billingHistory: BillingRecord[] = [];

    if (!isEmpty(sortedStats)) {
        Object.entries(sortedStats).forEach(([ts, stats]) => {
            const billingRecordIndex = billingHistory.findIndex((record) =>
                isEqual(record.date, stripTimeFromDate(ts))
            );

            const taskCount = stats.length;
            const dataVolume = evaluateDataVolume(stats);
            const totalCost = evaluateTotalCost(dataVolume, taskCount);

            if (billingRecordIndex === -1) {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    getInitialBillingRecord(ts);

                billingHistory.push({
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                });
            } else {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    billingHistory[billingRecordIndex];

                billingHistory[billingRecordIndex] = {
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                };
            }
        });
    }

    return billingHistory;
};

export interface SeriesConfig {
    data: [string, number][];
    seriesName?: string;
    stack?: string;
}

export const formatDataVolumeForDisplay = (
    seriesConfigs: SeriesConfig[],
    tooltipConfig: any
): string => {
    const filteredSeries =
        seriesConfigs.length === 1
            ? seriesConfigs
            : seriesConfigs.filter(
                  (series) => series.seriesName === tooltipConfig.seriesName
              );

    const dataVolumeInBytes = filteredSeries
        .flatMap(({ data }) => data)
        .find(([month]) => month === tooltipConfig.name);

    return dataVolumeInBytes
        ? prettyBytes(dataVolumeInBytes[1], { minimumFractionDigits: 2 })
        : `${tooltipConfig.value[1]} GB`;
};
