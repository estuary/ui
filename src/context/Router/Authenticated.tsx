import AppLayout from 'app/Layout';
import { AuthenticatedOnlyContext } from 'context/Authenticated';
import AuthenticatedHydrators from 'context/AuthenticatedHydrators';
import MarketplaceVerification from 'context/MarketplaceVerification';
import OnLoadSpinner from 'context/OnLoadSpinner/OnLoadSpinner';
import PaymentMethodWarning from 'context/PaymentMethodWarning';

function Authenticated() {
    return (
        <AuthenticatedOnlyContext>
            <AuthenticatedHydrators>
                <OnLoadSpinner display={false}>
                    <MarketplaceVerification>
                        <PaymentMethodWarning>
                            <AppLayout />
                        </PaymentMethodWarning>
                    </MarketplaceVerification>
                </OnLoadSpinner>
            </AuthenticatedHydrators>
        </AuthenticatedOnlyContext>
    );
}

export default Authenticated;
