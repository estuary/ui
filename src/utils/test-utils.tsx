/* eslint-disable @typescript-eslint/unbound-method */
// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { AuthContext } from 'context/Auth';
import AppContent from 'context/Content';
import AppRouter from 'context/Router';
import AppTheme from 'context/Theme';
import { ReactElement } from 'react';
import { auth, AuthDetails } from 'services/auth';

// TODO - this does not really do anything because we hardcode
//   the username down below on the MockAuthProvider. However,
//   this will be needed in the future so leaving the work so far
//   so we can leverage this when wanting to test the "on load auth"
//   functionality
const loginAsUser = (username: string) => {
    const mockAuthDetails: AuthDetails = {
        session: {
            account_id: '',
            expires_at: '',
            token: '',
        },
        user: {
            created_at: '',
            display_name: username,
            email: `${username}@${username}`,
            id: '',
            name: username,
            unique_name: username,
            updated_at: '',
        },
    };

    auth.saveSession(mockAuthDetails.session);
    auth.saveUser(mockAuthDetails.user);
};

const logoutUser = () => {
    auth.removeAuthDetails();
};

const goTo = (route?: string, name?: string) => {
    window.history.pushState(
        {},
        name ? name : 'Test page',
        route ? route : '/'
    );
};

const waitForLoading = async () => {
    return Promise.resolve();
};

// Mocking AuthContext out basically skips all the loading screen and boostrapping
//  stuff that happens on load. This should be fine for unit testing as it
//  can help speed up tests a bit and makes things easier to deal with. This
//  does mean that we need to eventually write a basic test that handles the
//  account bootstrapping.
interface MockAuthProps {
    children: JSX.Element;
    username?: string;
}
const MockAuthProvider = ({ children, username }: MockAuthProps) => {
    const value = {
        login: jest.fn(() => {
            if (username) {
                loginAsUser(username);
            }
            return Promise.resolve();
        }),
        logout: jest.fn(() => {
            logoutUser();
            return Promise.resolve();
        }),
        user: username,
    };

    return (
        <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
};

const customRender = async (
    ui: ReactElement,
    options: Omit<RenderOptions, 'wrapper'> & {
        route?: string;
        user?: string;
    }
) => {
    const { route, user } = options;

    goTo(route, 'Test Page');

    const view = rtlRender(
        <AppContent>
            <MockAuthProvider username={user}>
                <AppTheme>
                    <AppRouter>{ui}</AppRouter>
                </AppTheme>
            </MockAuthProvider>
        </AppContent>,
        {
            ...options,
        }
    );

    await waitForLoading();

    return view;
};

export * from '@testing-library/react';
export { goTo, loginAsUser, logoutUser, customRender, waitForLoading };
