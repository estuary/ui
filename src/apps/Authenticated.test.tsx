import '@testing-library/jest-dom';
import { customRender } from '../utils/test-utils';
import Authenticated from './Authenticated';

test('Authenticated app renders correctly', async () => {
    const { container } = await customRender(<Authenticated />, {
        user: 'FooBar123',
    });

    expect(container).toMatchSnapshot();
});
