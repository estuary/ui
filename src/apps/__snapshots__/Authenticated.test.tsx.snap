// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticated app renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-lglgsn"
  >
    <div
      class="MuiBox-root css-175rvmj"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1mifq5q-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-dv03od-MuiToolbar-root"
        >
          <div
            class="MuiBox-root css-12z0wuy"
          >
            <button
              aria-label="Expand Navigation"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1aqok8w-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
            </button>
          </div>
          <div
            class="MuiBox-root css-zdpt2t"
          >
            <div
              class="MuiBox-root css-ji4awn"
            >
              <img
                alt="Estuary"
                src="logo-estuary.png"
                width="120"
              />
            </div>
          </div>
          <div
            class="css-1f39f1a-MuiStack-root"
          >
            <button
              aria-controls="help-menu-menu"
              aria-haspopup="true"
              aria-label="Open help"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-j8rmhe-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              id="help-menu-button"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="HelpIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
                />
              </svg>
            </button>
            <button
              aria-controls="account-menu-menu"
              aria-haspopup="true"
              aria-label="Open account menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-j8rmhe-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              id="account-menu-button"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-12nwhkj-MuiAvatar-root"
              >
                T
              </div>
            </button>
          </div>
        </div>
      </header>
    </div>
    <div
      class="MuiBox-root css-8ujnwz"
    >
      <div
        class="MuiDrawer-root MuiDrawer-docked css-bippya-MuiDrawer-docked"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-12i7wg6-MuiPaper-root-MuiDrawer-paper"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-pqgv1q-MuiToolbar-root"
          />
          <div
            class="MuiBox-root css-1kzo3b9"
          >
            <ul
              aria-label="main application navigation"
              class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
            >
              <li>
                <a
                  aria-label="Collections"
                  class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root  css-gjl4xh-MuiButtonBase-root-MuiListItemButton-root Mui-disabled"
                  data-mui-internal-clone-element="true"
                  href="/app/collections"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ExploreIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 10.9c-.61 0-1.1.49-1.1 1.1s.49 1.1 1.1 1.1c.61 0 1.1-.49 1.1-1.1s-.49-1.1-1.1-1.1zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm2.19 12.19L6 18l3.81-8.19L18 6l-3.81 8.19z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1ler6va-MuiTypography-root"
                    >
                      Collections
                    </span>
                  </div>
                </a>
              </li>
              <li>
                <a
                  aria-label="Captures"
                  class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root  css-gjl4xh-MuiButtonBase-root-MuiListItemButton-root"
                  data-mui-internal-clone-element="true"
                  href="/app/captures"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="InputIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M21 3.01H3c-1.1 0-2 .9-2 2V9h2V4.99h18v14.03H3V15H1v4.01c0 1.1.9 1.98 2 1.98h18c1.1 0 2-.88 2-1.98v-14c0-1.11-.9-2-2-2zM11 16l4-4-4-4v3H1v2h10v3z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1ler6va-MuiTypography-root"
                    >
                      Captures
                    </span>
                  </div>
                </a>
              </li>
              <li>
                <a
                  aria-label="Materializations"
                  class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root  css-gjl4xh-MuiButtonBase-root-MuiListItemButton-root Mui-disabled"
                  data-mui-internal-clone-element="true"
                  href="/app/materializations"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="StorageIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M2 20h20v-4H2v4zm2-3h2v2H4v-2zM2 4v4h20V4H2zm4 3H4V5h2v2zm-4 7h20v-4H2v4zm2-3h2v2H4v-2z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1ler6va-MuiTypography-root"
                    >
                      Materializations
                    </span>
                  </div>
                </a>
              </li>
              <li>
                <a
                  aria-label="Derivations"
                  class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root  css-gjl4xh-MuiButtonBase-root-MuiListItemButton-root Mui-disabled"
                  data-mui-internal-clone-element="true"
                  href="/app/derivations"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CodeIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M9.4 16.6 4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0 4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1ler6va-MuiTypography-root"
                    >
                      Derivations
                    </span>
                  </div>
                </a>
              </li>
              <li>
                <a
                  aria-label="Change Set"
                  class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root  css-gjl4xh-MuiButtonBase-root-MuiListItemButton-root"
                  data-mui-internal-clone-element="true"
                  href="/app/change-set"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CompareArrowsIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M9.01 14H2v2h7.01v3L13 15l-3.99-4v3zm5.98-1v-3H22V8h-7.01V5L11 9l3.99 4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1ler6va-MuiTypography-root"
                    >
                      Change Set
                    </span>
                  </div>
                </a>
              </li>
              <li>
                <a
                  aria-label="Administration"
                  class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root  css-gjl4xh-MuiButtonBase-root-MuiListItemButton-root"
                  data-mui-internal-clone-element="true"
                  href="/app/admin"
                  role="button"
                  tabindex="0"
                >
                  <div
                    class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="HomeRepairServiceIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18 16h-2v-1H8v1H6v-1H2v5h20v-5h-4zm2-8h-3V6c0-1.1-.9-2-2-2H9c-1.1 0-2 .9-2 2v2H4c-1.1 0-2 .9-2 2v4h4v-2h2v2h8v-2h2v2h4v-4c0-1.1-.9-2-2-2zm-5 0H9V6h6v2z"
                      />
                    </svg>
                  </div>
                  <div
                    class="MuiListItemText-root css-tlelie-MuiListItemText-root"
                  >
                    <span
                      class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1ler6va-MuiTypography-root"
                    >
                      Administration
                    </span>
                  </div>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-b8j5vm"
    >
      <div
        class="MuiContainer-root css-u86gcf-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-pqgv1q-MuiToolbar-root"
        />
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-1hizfiu-MuiPaper-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-18qsmgm-MuiToolbar-root"
          >
            <h3
              class="MuiTypography-root MuiTypography-h3 css-1l7gydo-MuiTypography-root"
            >
              Welcome to Control Plane!
            </h3>
          </div>
          <div
            class="MuiBox-root css-1dedz41"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1pkqv6n-MuiTypography-root"
            >
              Click the Capture link over on the side navigation to get started.
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
